pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java

```diff
@@ -373,13 +372,31 @@ public void reloadSegment(String segmentName, IndexLoadingConfig indexLoadingCon
       } else {
         LOGGER.info("Reload existing segment: {} of table: {} on tier: {}", segmentName, _tableNameWithType,
             TierConfigUtils.normalizeTierName(segmentTier));
+        SegmentDirectory segmentDirectory =
+            initSegmentDirectory(segmentName, String.valueOf(zkMetadata.getCrc()), indexLoadingConfig);
+        if ((zkMetadata.getTier().equals(segmentTier))) {
+          // We should first try to reuse existing segment directory
+          boolean needReprocess = ImmutableSegmentLoader.needPreprocess(segmentDirectory, indexLoadingConfig, schema);
+          if (!needReprocess) {
+            LOGGER.info(
+                "Reloading segment : {} of table : {} using existing segment directory as no reprocessing needed",
+                segmentName, _tableNameWithType);
+            // No reprocessing needed, reuse the same segment
+            ImmutableSegment segment = ImmutableSegmentLoader.load(segmentDirectory, indexLoadingConfig, schema);
+            addSegment(segment);
+            return;
+          } else {
+            LOGGER.info("Segment : {} of table : {} requires reprocessing before reloading", segmentName,
+                _tableNameWithType);
+          }
+        }
+        // Create backup directory to handle failure of segment reloading.
+        createBackup(indexDir);
         // The indexDir is empty after calling createBackup, as it's renamed to a backup directory.
         // The SegmentDirectory should initialize accordingly. Like for SegmentLocalFSDirectory, it
         // doesn't load anything from an empty indexDir, but gets the info to complete the copyTo.
-        try (SegmentDirectory segmentDirectory = initSegmentDirectory(segmentName, String.valueOf(zkMetadata.getCrc()),
-            indexLoadingConfig)) {
-          segmentDirectory.copyTo(indexDir);
-        }
+        segmentDirectory.copyTo(indexDir);
```

Comment:
to be safe, better do `try { copyTo() } finally {close()}`

Reply:
Ack